criu/pagemap-cache.c

```diff
@@ -149,39 +160,75 @@ static int pmc_fill_cache(pmc_t *pmc, const struct vma_area *vma)
 			pr_debug("\t%d: simple mode [l:%lx h:%lx]\n", pmc->pid, pmc->start, pmc->end);
 	}
 
+	return pmc_fill(pmc, pmc->start, pmc->end);
+}
+
+int pmc_fill(pmc_t *pmc, u64 start, u64 end)
+{
+	size_t size_map;
+
+	pmc->start = start;
+	pmc->end = end;
+
 	size_map = PAGEMAP_LEN(pmc->end - pmc->start);
 	BUG_ON(pmc->map_len < size_map);
 	BUG_ON(pmc->fd < 0);
 
-	if (pread(pmc->fd, pmc->map, size_map, PAGEMAP_PFN_OFF(pmc->start)) != size_map) {
-		pmc_zap(pmc);
-		pr_perror("Can't read %d's pagemap file", pmc->pid);
-		return -1;
+	if (pmc->regs) {
+		struct pm_scan_arg args = {
+			.size = sizeof(struct pm_scan_arg),
+			.flags = 0,
+			.start = pmc->start,
+			.end = pmc->end,
+			.vec = (long)pmc->regs,
+			.vec_len = pmc->regs_max_len,
+			.max_pages = 0,
+			.category_inverted = PAGE_IS_PFNZERO | PAGE_IS_FILE,
+			.category_mask = PAGE_IS_PFNZERO | PAGE_IS_FILE,
+			.category_anyof_mask = PAGE_IS_PRESENT | PAGE_IS_SWAPPED,
+			.return_mask = PAGE_IS_PRESENT | PAGE_IS_SWAPPED | PAGE_IS_SOFT_DIRTY,
```

Comment:
Adding a comment to explain what exactly are we querying/filtering for might be nice here.